examples/raja-teams.cpp

```diff
@@ -0,0 +1,178 @@
+    RAJA::expt::omp_launch_t
+#else
+    RAJA::expt::seq_launch_t
+#endif
+#if defined(RAJA_ENABLE_CUDA)
+    ,
+    RAJA::expt::cuda_launch_t<false>
+#endif
+#if defined(RAJA_ENABLE_HIP)
+    ,
+    RAJA::expt::hip_launch_t<false>
+#endif
+    >;
+
+/*
+ * Define team policies.
+ * Up to 3 dimension are supported: x,y,z
+ */
+using teams_x = RAJA::expt::LoopPolicy<RAJA::loop_exec
+#if defined(RAJA_ENABLE_CUDA)
+                                       ,
+                                       RAJA::cuda_block_x_direct
+#endif
+#if defined(RAJA_ENABLE_HIP)
+                                       ,
+                                       RAJA::hip_block_x_direct
+#endif
+                                       >;
+/*
+ * Define thread policies.
+ * Up to 3 dimension are supported: x,y,z
+ */
+using threads_x = RAJA::expt::LoopPolicy<RAJA::loop_exec
+#if defined(RAJA_ENABLE_CUDA)
+                                         ,
+                                         RAJA::cuda_thread_x_loop
+#endif
+#if defined(RAJA_ENABLE_HIP)
+                                         ,
+                                         RAJA::hip_thread_x_loop
+#endif
+                                         >;
+
+
+int main(int RAJA_UNUSED_ARG(argc), char **RAJA_UNUSED_ARG(argv[]))
+{
+
+  // Resource object for host
+  camp::resources::Host host_res;
+
+  // Resource objects for CUDA or HIP
+#if defined(RAJA_ENABLE_CUDA)
+  camp::resources::Cuda device_res;
+#endif
+
+#if defined(RAJA_ENABLE_HIP)
+  camp::resources::Hip device_res;
+#endif
+
+  std::cout << "\n Running RAJA-Teams examples...\n";
+
+  for (int exec_place = 0; exec_place < (int)RAJA::expt::NUM_PLACES;
+       ++exec_place) {
+    RAJA::expt::ExecPlace select_cpu_or_gpu = (RAJA::expt::ExecPlace)exec_place;
+
+    // auto select_cpu_or_gpu = RAJA::HOST;
+    // auto select_cpu_or_gpu = RAJA::DEVICE;
+
+    int N_tri = 5;
+    int *Ddat;
+    if (select_cpu_or_gpu == RAJA::expt::HOST)
+      Ddat = host_res.allocate<int>(N_tri * N_tri);
+
+#if defined(RAJA_ENABLE_DEVICE)
+    if (select_cpu_or_gpu == RAJA::expt::DEVICE)
+      Ddat = device_res.allocate<int>(N_tri * N_tri);
+#endif
+
+    /*
+     * launch just starts a "kernel" it's doesn't provide any looping.
+     *
+     * The first argument determines which policy should be executed,
+     *
+     * The second argument is the number of teams+threads needed for each of the
+     * policies.
+     *
+     * Third argument is the lambda for the policy.
+     *
+     *
+     * The lambda takes a "resource" object, which has the teams+threads
+     */
+
+
+    std::cout << "\n Running Upper triangular pattern example...\n";
+
+
+    RAJA::View<int, RAJA::Layout<2>> D(Ddat, N_tri, N_tri);
+
+    RAJA::expt::launch<launch_policy>(
+        select_cpu_or_gpu,
+        RAJA::expt::Resources(RAJA::expt::Teams(N_tri),
+                              RAJA::expt::Threads(N_tri)),
+        [=] RAJA_HOST_DEVICE(RAJA::expt::LaunchContext ctx) {
```

Comment:
Everything looks good overall. This may be for a future PR, but it seems like the LaunchContext can encapsulate Teams() and Threads(). The LaunchContext gets passed all the way into launch() and we could probably extract more use from it beyond the ctx.teamSync() a few lines below.

Reply:
Yeah, I think that would make substantially more sense.  Something we can explore in the trip to stabilizing it probably.  I'm going to extract that comment into an issue so we can track it.